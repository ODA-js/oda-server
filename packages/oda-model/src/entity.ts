import {
  EntityBasePersistence,
  EntityBaseInternal,
  EntityBaseMetaInfo,
  EntityBaseInput,
  IEntityBase,
  EntityBase,
  EntityBaseOutput,
} from './entitybase';
import { MetaModelType, Nullable, assignValue } from './types';
import { merge } from 'lodash';

export interface IEntity
  extends IEntityBase<
    EntityMetaInfo,
    EntityPersistence,
    EntityInput,
    EntityOutput
  > {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityInternal
  extends EntityBaseInternal<EntityMetaInfo, EntityPersistence> {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityPersistence extends EntityBasePersistence {}

export interface EntityInput
  extends EntityBaseInput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityOutput
  extends EntityBaseOutput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityMetaInfo extends EntityBaseMetaInfo<EntityPersistence> {}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Entity
  extends EntityBase<
    EntityMetaInfo,
    EntityInput,
    EntityInternal,
    EntityPersistence,
    EntityOutput
  >
  implements IEntity {
  public get modelType(): MetaModelType {
    return 'entity';
  }

  constructor(init: EntityInput) {
    super(merge({}, defaultInput, init));
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return this.$obj.abstract;
  }

  get embedded(): boolean | Set<string> {
    return this.$obj.embedded;
  }

  public updateWith(input: Nullable<EntityInput>) {
    super.updateWith(input);

    assignValue({
      src: this.$obj,
      input,
      field: 'abstract',
      setDefault: src => (src.abstract = false),
    });
    assignValue<EntityInternal, EntityInput, string[]>({
      src: this.$obj,
      input,
      field: 'implements',
      effect: (src, value) => (src.implements = new Set(value)),
      setDefault: src => (src.implements = new Set()),
    });

    assignValue<EntityInternal, EntityInput, boolean | string[]>({
      src: this.$obj,
      input,
      field: 'embedded',
      effect: (src, value) =>
        (src.embedded = typeof value === 'boolean' ? value : new Set(value)),
      setDefault: src => (src.embedded = false),
    });
  }

  public toObject(): EntityOutput {
    return merge({}, super.toObject(), {
      implements: [...this.implements.values()],
      embedded: this.embedded,
      abstract: this.abstract,
    } as Partial<EntityOutput>);
  }
}
