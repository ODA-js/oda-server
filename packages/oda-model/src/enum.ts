import {
  ModelBase,
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBaseOutput,
} from './modelbase';
import { EnumItem, IEnumItem, EnumItemInput } from './enumItem';
import { merge } from 'lodash';
import { ElementMetaInfo } from './element';
import { MetaModelType, Nullable, assignValue } from './types';

export interface IEnum extends IModelBase<EnumMetaInfo, EnumInput, EnumOutput> {
  /**
   * Enum item definition
   */
  readonly items: Map<string, IEnumItem>;
}

export interface EnumMetaInfo extends ElementMetaInfo {}

export interface EnumInternal extends ModelBaseInternal<EnumMetaInfo> {
  items: Map<string, IEnumItem>;
}

export interface EnumInput extends ModelBaseInput<EnumMetaInfo> {
  items: (EnumItemInput | string)[];
}

export interface EnumOutput extends ModelBaseOutput<EnumMetaInfo> {
  items: EnumItemInput[];
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Enum
  extends ModelBase<EnumMetaInfo, EnumInput, EnumInternal, EnumOutput>
  implements IEnum {
  public get modelType(): MetaModelType {
    return 'enum';
  }
  get items() {
    return this.$obj.items;
  }

  constructor(init: EnumInput) {
    super(merge({}, defaultInput, init));
  }

  public updateWith(input: Nullable<EnumInput>) {
    super.updateWith(input);

    assignValue<EnumInternal, EnumInput, EnumInput['items']>({
      src: this.$obj,
      input,
      field: 'items',
      effect: (src, value) =>
        (src.items = new Map(
          value
            .map(i =>
              typeof i === 'string' ? ({ name: i } as EnumItemInput) : i,
            )
            .map(
              i =>
                [
                  i.name,
                  new EnumItem({
                    name: i.name,
                    title: i.title || i.name,
                    description: i.description || i.title || i.name,
                    value: i.value,
                    metadata: i.metadata,
                  }),
                ] as [string, IEnumItem],
            ),
        )),
      required: true,
    });
  }

  public toObject(): EnumOutput {
    return merge({}, super.toObject(), {
      items: [...this.$obj.items.values()].map(i => i.toObject()),
    } as Partial<EnumOutput>);
  }
}
