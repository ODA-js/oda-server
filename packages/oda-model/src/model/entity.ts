import clean from '../lib/json/clean';
import { Field } from './field';
import {
  EntityInput,
  EntityJSON,
  EntityStorage,
  IEntity,
  MetaModelType,
} from './interfaces';
import { ModelPackage } from './modelpackage';
import { EntityBase } from './entitybase';

/**
 * 1. тип объекта который входит на updateWith
 * 2. тип объекта который идет на toObject
 * 3. тип объекта который идет на toJSON
 * 3. тип объекта который идет на выходе clone
 */

export class Entity extends EntityBase implements IEntity {
  public modelType: MetaModelType = 'entity';

  protected $obj!: EntityStorage;

  constructor(obj: EntityInput) {
    super(obj);
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return !!this.$obj.abstract;
  }

  get embedded(): boolean | string[] {
    return this.$obj.embedded instanceof Set
      ? Array.from(this.$obj.embedded)
      : this.$obj.embedded;
  }

  public ensureImplementation(modelPackage: ModelPackage) {
    const newFields: Map<string, Field> = new Map<string, Field>();
    this.implements.forEach(intrf => {
      if (modelPackage.mixins.has(intrf)) {
        const impl = modelPackage.mixins.get(intrf);
        impl.fields.forEach(f => {
          if (!this.fields.has(f.name)) {
            newFields.set(f.name, f);
          }
        });
      }
    });

    if (newFields.size > 0) {
      const update = this.toJSON();
      update.fields.push(...[...newFields.values()].map(f => f.toJSON()));
      this.updateWith(update);
      this.ensureIds(modelPackage);
    }
  }

  public updateWith(obj: EntityInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };
      const impl = new Set(obj.implements);
      result.embedded = Array.isArray(obj.embedded)
        ? new Set(obj.embedded)
        : obj.embedded;
      result.abstract = obj.abstract;
      result.implements = impl;
      this.$obj = result;
    }
  }

  public toObject(modelPackage?: ModelPackage) {
    if (!modelPackage) {
      let res = super.toObject();
      return clean({
        ...res,
        implements: [...this.implements],
        embedded: this.embedded,
        abstract: this.abstract,
      });
    } else {
      let modelRelations = modelPackage.relations.get(this.name);
      if (modelRelations) {
        let res = super.toObject();
        return clean({
          ...res,
          embedded: this.embedded,
          abstract: this.abstract,
          implements: [...this.implements].filter(i =>
            modelPackage.mixins.has(i),
          ),
        });
      }
    }
  }
}
