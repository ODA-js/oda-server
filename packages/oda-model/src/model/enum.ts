import clean from '../lib/json/clean';
import { MetaModelType, EnumStorage, EnumInput } from './interfaces';
import { ModelBase } from './modelbase';
import { IEnum, IEnumItemsInput } from './interfaces/model';

export class Enum extends ModelBase implements IEnum {
  public modelType: MetaModelType = 'union';
  protected $obj!: EnumStorage;

  get items(): IEnumItemsInput[] {
    return this.$obj.items;
  }

  public updateWith(obj: EnumInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };
      result.name = obj.name;
      let $items = obj.items;
      let items = $items;

      result.items = items
        .map(i => (typeof i === 'string' ? { name: i } : i))
        .map(i => ({
          name: i.name,
          title: i.title || i.name,
          description: i.description || i.title || i.name,
          value: i.value,
          metadata: i.metadata,
        }));

      result.items_ = $items;

      this.$obj = result;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      items: props.items,
    });
  }
}
