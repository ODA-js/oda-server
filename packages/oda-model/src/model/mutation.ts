import {
  MutationInput,
  MutationStorage,
  MutationMeta,
  MapToHash,
  HashToMap,
  MetaModelType,
} from './interfaces';
import { ModelBase } from './modelbase';
import { IMutation } from './interfaces/model';

export class Mutation
  extends ModelBase<MutationMeta, MutationInput, MutationStorage>
  implements IMutation {
  public modelType: MetaModelType = 'mutation';
  constructor(inp: MutationInput) {
    super(inp);
  }
  public get args() {
    return this.$obj.args;
  }

  public get payload() {
    return this.$obj.payload;
  }

  public updateWith(obj: MutationInput) {
    if (obj) {
      super.updateWith(obj);
      this.$obj.args = HashToMap(obj.args);
      this.$obj.payload = HashToMap(obj.payload);
    }
  }

  public toObject(): MutationInput {
    return {
      ...super.toObject(),
      args: MapToHash(this.$obj.args),
      payload: MapToHash(this.$obj.payload),
    };
  }
}
