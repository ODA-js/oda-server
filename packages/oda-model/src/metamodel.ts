import * as fs from 'fs';

import { Entity, EntityInput } from './entity';

import {
  ModelPackage,
  IPackage,
  PackageMetaInfo,
  ModelPackageInternal,
  ModelPackageInput,
} from './modelpackage';
import { Mutation, MutationInput } from './mutation';
import { Query, QueryInput } from './query';
import { Mixin, MixinInput } from './mixin';
import { Union, UnionInput } from './union';
import { Enum, EnumInput } from './enum';
import { Scalar, ScalarInput } from './scalar';
import { Directive, DirectiveInput } from './directive';
import { MetaModelType, AsHash } from './model';
import { IModelBase, ModelBaseInput } from './modelbase';
import fold from './lib/json/fold';

function getFilter(inp: string): { filter: (f) => boolean; fields: string[] } {
  let result = {
    filter: f => f.name === inp,
    fields: [inp],
  };
  if (inp === '*') {
    result.filter = () => true;
  }
  if (inp.startsWith('^[')) {
    let notFields = inp
      .slice(2, inp.length - 1)
      .split(',')
      .map(f => f.trim())
      .reduce((res, cur) => {
        res[cur] = 1;
        return res;
      }, {});
    result.filter = f => !notFields[f.name];
    result.fields = [];
  }
  if (inp.startsWith('[')) {
    let onlyFields = inp
      .slice(1, inp.length - 1)
      .split(',')
      .map(f => f.trim())
      .reduce((res, cur) => {
        res[cur] = 1;
        return res;
      }, {});
    result.filter = f => !!onlyFields[f.name];
    result.fields = Object.keys(onlyFields);
  }
  return result;
}

export interface IModelHook {
  name: string;
  entities?: AsHash<EntityInput>;
  mutations?: AsHash<MutationInput>;
  queries?: AsHash<QueryInput>;
}

export interface IModel extends IModelBase<MetaModelMetaInfo, MetaModelInput> {
  readonly packages: Map<string, IPackage>;
}

export interface MetaModelInput extends ModelBaseInput<MetaModelMetaInfo> {
  entities: EntityInput[];
  packages: ModelPackageInput[];
  mutations?: MutationInput[];
  queries?: QueryInput[];
  scalars: ScalarInput[];
  directives: DirectiveInput[];
  enums?: EnumInput[];
  unions?: UnionInput[];
  mixins?: MixinInput[];
  name: string;
  title?: string;
  description?: string;
}

export interface MetaModelMetaInfo extends PackageMetaInfo {}

export interface MetaModelInternal extends ModelPackageInternal {}

/**
 * Represents meta-model store
 */
export class MetaModel extends ModelPackage implements IModel {
  public modelType: MetaModelType = 'model';
  public packages: Map<string, ModelPackage> = new Map();
  public store: string = 'default.json';
  public defaultPackage: ModelPackage;

  constructor(name: string = 'default') {
    super(name);
    this.ensureDefaultPackage();
  }

  public loadModel(fileName: string = this.store) {
    let txt = fs.readFileSync(fileName);
    let store = JSON.parse(txt.toString()) as MetaModelInput;
    this.loadPackage(store);
  }

  protected dedupeFields(
    src: SimpleFieldInput[],
  ): { [key: string]: SimpleFieldInput } {
    return src.reduce(
      (res, curr) => {
        if (!res.hasOwnProperty(curr.name)) {
          res[curr.name] = curr;
        } else {
          res[curr.name] = deepMerge(res[curr.name], curr);
        }
        return res;
      },
      {} as { [key: string]: SimpleFieldInput },
    );
  }

  protected applyEntityHook(entity: Entity, hook: EntityInput): Entity {
    let result = entity.toJSON();
    let metadata;
    if (hook.metadata) {
      metadata = deepMerge(result.metadata || {}, hook.metadata);
    }
    let fields: {
      [fName: string]: SimpleFieldInput;
    };

    if (Array.isArray(hook.fields)) {
      fields = this.dedupeFields([
        ...(result.fields as SimpleFieldInput[]),
        ...(hook.fields as SimpleFieldInput[]),
      ]);
    } else {
      fields = this.dedupeFields(result.fields || []);
      let fNames = Object.keys(hook.fields || {});
      for (let i = 0, len = fNames.length; i < len; i++) {
        let fName = fNames[i];
        let prepare = getFilter(fName);
        let list = (result.fields || []).filter(prepare.filter);
        if (list.length > 0 && hook.fields !== undefined) {
          list.forEach(f => {
            if (hook.fields) {
              fields[f.name] = deepMerge(f, hook.fields[fName] as any);
            }
          });
        } else {
          // create specific items
          prepare.fields.forEach(f => {
            fields[f] = hook.fields[fName];
          });
        }
      }
    }
    return new Entity({
      ...result,
      fields: fields !== undefined ? fields : undefined,
      metadata,
    });
  }

  protected applyMutationHook(
    mutation: Mutation,
    hook: MutationInput,
  ): Mutation {
    let result = mutation.toObject();
    let metadata;
    if (hook.metadata) {
      metadata = deepMerge(result.metadata || {}, hook.metadata);
    }

    let args = result.args,
      payload = result.payload;
    if (hook.args) {
      args = [...args, ...hook.args];
    }

    if (hook.payload) {
      payload = [...payload, ...hook.payload];
    }

    result = {
      ...result,
      args,
      payload,
      metadata,
    };
    return new Mutation(result);
  }

  protected applyQueryHook(mutation: Query, hook: QueryInput): Query {
    let result = mutation.toJSON() as MutationInput;
    let metadata;
    if (hook.metadata) {
      metadata = deepMerge(result.metadata || {}, hook.metadata);
    }

    let args = result.args,
      payload = result.payload;
    if (hook.args) {
      args = [...args, ...hook.args];
    }

    if (hook.payload) {
      payload = [...payload, ...hook.payload];
    }

    result = {
      ...result,
      args,
      payload,
      metadata,
    };
    return new Query(result);
  }

  public applyHooks(hooks?: IModelHook[]) {
    if (hooks && !Array.isArray(hooks)) {
      hooks = [hooks];
    }
    if (hooks) {
      hooks = hooks.filter(f => f);
      hooks.forEach(hook => {
        if (hook.entities && this.entities.size > 0) {
          let keys = Object.keys(hook.entities);
          for (let i = 0, len = keys.length; i < len; i++) {
            let key = keys[i];
            let current = hook.entities[key];
            current.fields = current.fields ? current.fields : [];
            current.metadata = current.metadata ? current.metadata : {};
            let prepare = getFilter(key);
            let list = Array.from(this.entities.values()).filter(
              prepare.filter,
            );
            if (list.length > 0) {
              list.forEach(e => {
                let result = this.applyEntityHook(e, current);
                this.entities.set(result.name, result);
              });
            } else if (prepare.fields.length > 0) {
              throw new Error(
                `Entit${prepare.fields.length === 1 ? 'y' : 'ies'} ${
                  prepare.fields
                } not found`,
              );
            }
          }
        }
        if (hook.mutations && this.mutations.size > 0) {
          let keys = Object.keys(hook.mutations);
          for (let i = 0, len = keys.length; i < len; i++) {
            let key = keys[i];
            let current = hook.mutations[key];
            current.args = current.args ? current.args : [];
            current.payload = current.payload ? current.payload : [];
            current.metadata = current.metadata ? current.metadata : {};

            let prepare = getFilter(key);
            let list = Array.from(this.mutations.values()).filter(
              prepare.filter,
            );
            if (list.length > 0) {
              list.forEach(e => {
                let result = this.applyMutationHook(e, current);
                this.mutations.set(result.name, result);
              });
            } else if (prepare.fields.length > 0) {
              throw new Error(
                `Mutation${prepare.fields.length > 1 ? 's' : ''} ${
                  prepare.fields
                } not found`,
              );
            }
          }
        }
        if (hook.queries && this.queries.size > 0) {
          let keys = Object.keys(hook.queries);
          for (let i = 0, len = keys.length; i < len; i++) {
            let key = keys[i];
            let current = hook.queries[key];
            current.args = current.args ? current.args : [];
            current.payload = current.payload ? current.payload : [];
            current.metadata = current.metadata ? current.metadata : {};

            let prepare = getFilter(key);
            let list = Array.from(this.queries.values()).filter(prepare.filter);
            if (list.length > 0) {
              list.forEach(e => {
                let result = this.applyQueryHook(e, current);
                this.queries.set(result.name, result);
              });
            } else if (prepare.fields.length > 0) {
              throw new Error(
                `Quer${prepare.fields.length > 1 ? 'ies' : 'y'} ${
                  prepare.fields
                } not found`,
              );
            }
          }
        }
      });
    }
  }

  public addPackage(pckg: ModelPackageInput) {
    let pack: ModelPackage;
    if (pckg.name && this.packages.has(pckg.name)) {
      pack = this.packages.get(pckg.name) as ModelPackage;
    } else {
      pack = new ModelPackage(pckg);
      pack.connect(this);
      this.packages.set(pckg.name, pack);
    }

    if (pack && pckg.mixins) {
      pckg.mixins.forEach(e => {
        if (this.mixins.has(e) && !pack.mixins.has(e)) {
          pack.addMixin(this.mixins.get(e));
        }
      });
    }

    if (pckg.entities) {
      pckg.entities.forEach(e => {
        if (this.entities.has(e) && !pack.entities.has(e)) {
          pack.addEntity(this.entities.get(e));
        }
      });
    }

    if (pckg.mutations) {
      pckg.mutations.forEach(m => {
        if (this.mutations.has(m) && !pack.mutations.has(m)) {
          pack.addMutation(this.mutations.get(m));
        }
      });
    }

    if (pckg.queries) {
      pckg.queries.forEach(e => {
        if (this.queries.has(e) && !pack.queries.has(e)) {
          pack.addQuery(this.queries.get(e));
        }
      });
    }

    if (pckg.enums) {
      pckg.enums.forEach(e => {
        if (this.enums.has(e) && !pack.enums.has(e)) {
          pack.addEnum(this.enums.get(e));
        }
      });
    }

    if (pckg.scalars) {
      pckg.scalars.forEach(e => {
        if (this.scalars.has(e) && !pack.scalars.has(e)) {
          pack.addScalar(this.scalars.get(e));
        }
      });
    }

    if (pckg.directives) {
      pckg.directives.forEach(e => {
        if (this.directives.has(e) && !pack.directives.has(e)) {
          pack.addDirective(this.directives.get(e));
        }
      });
    }

    if (pckg.unions) {
      pckg.unions.forEach(e => {
        if (this.unions.has(e) && !pack.unions.has(e)) {
          pack.addUnion(this.unions.get(e));
        }
      });
    }

    pack.ensureAll();
  }

  public loadPackage(store: MetaModelInput, hooks?: any[]) {
    this.reset();

    // must go first
    if (store.mixins) {
      store.mixins.forEach(q => {
        this.addMixin(new Mixin(q));
      });
    }

    if (store.entities) {
      store.entities.forEach(ent => {
        this.addEntity(new Entity(ent));
      });
    }
    if (store.mutations) {
      store.mutations.forEach(mut => {
        this.addMutation(new Mutation(mut));
      });
    }

    if (store.queries) {
      store.queries.forEach(q => {
        this.addQuery(new Query(q));
      });
    }

    if (store.enums) {
      store.enums.forEach(q => {
        this.addEnum(new Enum(q));
      });
    }

    if (store.scalars) {
      store.scalars.forEach(q => {
        this.addScalar(new Scalar(q));
      });
    }

    if (store.directives) {
      store.directives.forEach(q => {
        this.addDirective(new Directive(q));
      });
    }

    if (store.unions) {
      store.unions.forEach(q => {
        this.addUnion(new Union(q));
      });
    }

    this.ensureDefaultPackage();

    this.applyHooks(fold(hooks) as IModelHook[]);

    if (Array.isArray(store.packages)) {
      store.packages.forEach(this.addPackage.bind(this));
    }
  }

  public saveModel(fileName: string = this.store) {
    fs.writeFileSync(
      fileName,
      JSON.stringify({
        entities: Array.from(this.entities.values()).map(f => f.toJSON()),
        packages: Array.from(
          this.packages.values(),
        ) /*.filter(p => p.name !== 'default')*/
          .map(f => f.toObject()),
        mutations: Array.from(this.mutations.values()).map(f => f.toJSON()),
        queries: Array.from(this.queries.values()).map(f => f.toJSON()),
        enums: Array.from(this.enums.values()).map(f => f.toJSON()),
        unions: Array.from(this.unions.values()).map(f => f.toJSON()),
        mixins: Array.from(this.mixins.values()).map(f => f.toJSON()),
        scalars: Array.from(this.scalars.values()).map(f => f.toJSON()),
        directives: Array.from(this.directives.values()).map(f => f.toJSON()),
      }),
    );
  }

  public reset() {
    this.entities.clear();
    this.packages.clear();
    this.mutations.clear();
    this.queries.clear();
    this.enums.clear();
    this.mixins.clear();
    this.unions.clear();
    this.scalars.clear();
    this.directives.clear();
  }

  public createPackage(name: string): ModelPackage {
    if (this.packages.has(name)) {
      throw new Error(`Package "${name}" already exists`);
    }
    let pack = new ModelPackage(name);
    this.packages.set(name, pack);
    pack.connect(this);
    return pack;
  }

  public assignEntityToPackage(input: { entity: string; package: string }) {
    let pack = this.packages.get(input.package);
    if (!pack) {
      throw new Error(`Package ${input.package} didn't exists`);
    }
    let ent = this.entities.get(input.entity);
    if (!ent) {
      throw new Error(`Package ${input.entity} didn't exists`);
    }
    pack.addEntity(ent);
    pack.ensureAll();
    return {
      package: pack,
      entity: ent,
    };
  }

  private ensureDefaultPackage() {
    if (!this.packages.has(this.name)) {
      this.defaultPackage = this;
      this.connect(this);
      this.ensureAll();
      this.packages.set(this.name, this);
    }
  }
}
