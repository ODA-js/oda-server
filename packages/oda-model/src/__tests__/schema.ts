import { MetaModelInput } from '../metamodel';
import { EntityInput, EntityMetaInfo } from '../entity';
import { MutationInput } from '../mutation';
import { EnumInput } from '../enum';
import { QueryInput } from '../query';
import { ModelPackageInput } from '../modelpackage';

const MediaFile: EntityInput = {
  name: 'MediaFile',
  fields: [
    {
      name: 'title',
    },
    {
      name: 'description',
      type: 'RichText',
    },
    {
      name: 'reference',
      type: { type: 'entity', name: 'Image', multiplicity: 'one' },
    },
    {
      name: 'catalog',
      type: { type: 'entity', name: 'Image', multiplicity: 'many' },
    },
  ],
  metadata: { UI: { listName: ['title'] } } as Partial<EntityMetaInfo>,
};

const Image: EntityInput = {
  implements: ['IFile'],
  name: 'Image',
  embedded: true,
  fields: [
    {
      name: 'width',
      type: 'int',
    },
    {
      name: 'height',
      type: 'int',
    },
  ],
};

const IFile: EntityInput = {
  abstract: true,
  name: 'IFile',
  fields: [
    { name: 'src', identity: true },
    { name: 'type' },
    { name: 'path' },
    { name: 'name' },
  ],
  metadata: { UI: { listName: ['src'] } } as Partial<EntityMetaInfo>,
};

const User: EntityInput = {
  name: 'User',
  description: 'Person role to be user identified in the system',
  fields: [
    { name: 'userName', identity: true },
    { name: 'password' },
    { name: 'isAdmin', type: 'Boolean' },
    { name: 'isSystem', type: 'Boolean' },
    { name: 'enabled', type: 'Boolean' },
    {
      name: 'socialNetworks',
      type: {
        type: 'entity',
        name: 'SocialNetwork',
        multiplicity: 'many',
      },
    },
    {
      name: 'emails',
      relation: {
        hasMany: 'Email#',
      },
    },
    {
      name: 'files',
      relation: {
        hasMany: 'MediaFile#',
      },
    },
  ],
  metadata: { UI: { listName: ['userName'] } } as Partial<EntityMetaInfo>,
};

const Login: MutationInput = {
  name: 'loginUser',
  description: 'make user login',
  args: [
    {
      name: 'userName',
      required: true,
    },
    {
      name: 'password',
      required: true,
    },
  ],
  payload: [
    {
      name: 'token',
    },
    {
      name: 'role',
    },
  ],
};

const EMail: EntityInput = {
  name: 'Email',
  fields: [
    {
      name: 'email',
      identity: true,
    },
    {
      name: 'type',
      type: { type: 'enum', name: 'CommunicationType', multiplicity: 'one' },
    },
    {
      name: 'user',
      indexed: true,
      relation: {
        belongsTo: 'User#',
        opposite: 'emails',
      },
    },
  ],
};

const CommunicationType: EnumInput = {
  name: 'CommunicationType',
  items: ['home', 'office', 'other'],
};

const SocialNetwork: EntityInput = {
  embedded: true,
  name: 'SocialNetwork',
  fields: [
    { name: 'account', identity: true },
    { name: 'url', indexed: true },
    {
      name: 'type',
      type: {
        type: 'enum',
        multiplicity: 'one',
        name: 'SocialNetworkType',
      },
    },
  ],
};

const SocialNetworkType: EnumInput = {
  name: 'SocialNetworkType',
  items: ['skype', 'vk', 'tweeter', 'odn', 'whatsApp'],
};

const UserInfo: QueryInput = {
  name: 'UserInfo',
  args: [],
  payload: [{ name: 'id' }, { name: 'name' }, { name: 'email' }],
};

const AdminPackage: ModelPackageInput = {
  name: 'Admin',
  entities: ['User', 'Email', 'SocialNetwork'],
};

const PublicPackage: ModelPackageInput = {
  name: 'public',
  mutations: [Login],
};

const UserPackage_User: EntityInput = {
  name: 'User',
  description: 'Person role to be user identified in the system',
  fields: [
    { name: 'userName', identity: true },
    { name: 'password' },
    {
      name: 'socialNetworks',
      type: {
        type: 'entity',
        name: 'SocialNetwork',
        multiplicity: 'many',
      },
    },
    {
      name: 'emails',
      relation: {
        hasMany: 'Email#',
      },
    },
    {
      name: 'files',
      relation: {
        hasMany: 'MediaFile#',
      },
    },
  ],
  metadata: { UI: { listName: ['userName'] } } as Partial<EntityMetaInfo>,
};

const UserPackage: ModelPackageInput = {
  name: 'owner',
  entities: [UserPackage_User, 'MediaFile'],
};

const schema: MetaModelInput = {
  name: 'Education',
  packages: [AdminPackage, PublicPackage, UserPackage],
  entities: [MediaFile, Image, IFile, User, EMail, SocialNetwork],
  mutations: [Login],
  queries: [UserInfo],
  enums: [CommunicationType, SocialNetworkType],
};

export default schema;

it('schema', () => {
  expect(schema).toMatchSnapshot('schema');
});
