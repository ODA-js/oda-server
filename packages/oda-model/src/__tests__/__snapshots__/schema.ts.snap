// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema: schema 1`] = `
Object {
  "entities": Array [
    Object {
      "fields": Array [
        Object {
          "name": "title",
        },
        Object {
          "name": "description",
          "type": "RichText",
        },
        Object {
          "name": "reference",
          "type": Object {
            "multiplicity": "one",
            "name": "Image",
            "type": "entity",
          },
        },
        Object {
          "name": "catalog",
          "type": Object {
            "multiplicity": "many",
            "name": "Image",
            "type": "entity",
          },
        },
      ],
      "metadata": Object {
        "UI": Object {
          "listName": Array [
            "title",
          ],
        },
      },
      "name": "MediaFile",
    },
    Object {
      "embedded": true,
      "fields": Array [
        Object {
          "name": "width",
          "type": "int",
        },
        Object {
          "name": "height",
          "type": "int",
        },
      ],
      "implements": Array [
        "IFile",
      ],
      "name": "Image",
    },
    Object {
      "abstract": true,
      "fields": Array [
        Object {
          "identity": true,
          "name": "src",
        },
        Object {
          "name": "type",
        },
        Object {
          "name": "path",
        },
        Object {
          "name": "name",
        },
      ],
      "metadata": Object {
        "UI": Object {
          "listName": Array [
            "src",
          ],
        },
      },
      "name": "IFile",
    },
    Object {
      "description": "Person role to be user identified in the system",
      "fields": Array [
        Object {
          "identity": true,
          "name": "userName",
        },
        Object {
          "name": "password",
        },
        Object {
          "name": "isAdmin",
          "type": "Boolean",
        },
        Object {
          "name": "isSystem",
          "type": "Boolean",
        },
        Object {
          "name": "enabled",
          "type": "Boolean",
        },
        Object {
          "name": "socialNetworks",
          "type": Object {
            "multiplicity": "many",
            "name": "SocialNetwork",
            "type": "entity",
          },
        },
        Object {
          "name": "emails",
          "relation": Object {
            "hasMany": "Email#",
          },
        },
        Object {
          "name": "files",
          "relation": Object {
            "hasMany": "MediaFile#",
          },
        },
      ],
      "metadata": Object {
        "UI": Object {
          "listName": Array [
            "userName",
          ],
        },
      },
      "name": "User",
    },
    Object {
      "fields": Array [
        Object {
          "identity": true,
          "name": "email",
        },
        Object {
          "name": "type",
          "type": Object {
            "multiplicity": "one",
            "name": "CommunicationType",
            "type": "enum",
          },
        },
        Object {
          "indexed": true,
          "name": "user",
          "relation": Object {
            "belongsTo": "User#",
            "opposite": "emails",
          },
        },
      ],
      "name": "Email",
    },
    Object {
      "embedded": true,
      "fields": Array [
        Object {
          "identity": true,
          "name": "account",
        },
        Object {
          "indexed": true,
          "name": "url",
        },
        Object {
          "name": "type",
          "type": Object {
            "multiplicity": "one",
            "name": "SocialNetworkType",
            "type": "enum",
          },
        },
      ],
      "name": "SocialNetwork",
    },
  ],
  "enums": Array [
    Object {
      "items": Array [
        "home",
        "office",
        "other",
      ],
      "name": "CommunicationType",
    },
    Object {
      "items": Array [
        "skype",
        "vk",
        "tweeter",
        "odn",
        "whatsApp",
      ],
      "name": "SocialNetworkType",
    },
  ],
  "mutations": Array [
    Object {
      "args": Array [
        Object {
          "name": "userName",
          "required": true,
        },
        Object {
          "name": "password",
          "required": true,
        },
      ],
      "description": "make user login",
      "name": "loginUser",
      "payload": Array [
        Object {
          "name": "token",
        },
        Object {
          "name": "role",
        },
      ],
    },
  ],
  "name": "Education",
  "packages": Array [
    Object {
      "entities": Array [
        "User",
        "Email",
        "SocialNetwork",
      ],
      "name": "Admin",
    },
    Object {
      "mutations": Array [
        Object {
          "args": Array [
            Object {
              "name": "userName",
              "required": true,
            },
            Object {
              "name": "password",
              "required": true,
            },
          ],
          "description": "make user login",
          "name": "loginUser",
          "payload": Array [
            Object {
              "name": "token",
            },
            Object {
              "name": "role",
            },
          ],
        },
      ],
      "name": "public",
    },
    Object {
      "entities": Array [
        Object {
          "description": "Person role to be user identified in the system",
          "fields": Array [
            Object {
              "identity": true,
              "name": "userName",
            },
            Object {
              "name": "password",
            },
            Object {
              "name": "socialNetworks",
              "type": Object {
                "multiplicity": "many",
                "name": "SocialNetwork",
                "type": "entity",
              },
            },
            Object {
              "name": "emails",
              "relation": Object {
                "hasMany": "Email#",
              },
            },
            Object {
              "name": "files",
              "relation": Object {
                "hasMany": "MediaFile#",
              },
            },
          ],
          "metadata": Object {
            "UI": Object {
              "listName": Array [
                "userName",
              ],
            },
          },
          "name": "User",
        },
        "MediaFile",
      ],
      "name": "owner",
    },
  ],
  "queries": Array [
    Object {
      "args": Array [],
      "name": "UserInfo",
      "payload": Array [
        Object {
          "name": "id",
        },
        Object {
          "name": "name",
        },
        Object {
          "name": "email",
        },
      ],
    },
  ],
}
`;
